### DISCUSSION TITLE
Query timeout  and error type is INSUFFICIENT_RESOURCES

### DISCUSSION THREAD

**Discussion by njalan** [trinodb/trino#23199] [Q&A] [answered]
In my current cluster each 15 minutes I can see some queries are timeout. But at least 95% queries are finished successfully and these failed queries are successfully when run again.
 
![image](https://github.com/user-attachments/assets/f3abc6d5-9383-4395-8b39-77ff49b34480)

**Comment by Heltman** [comment, marked as answer]
What is the size of your cluster, how is it configured, what is the concurrency of queries, and what is the size distribution of queries? 

Generally speaking, when the cluster pressure is high, the query time will increase and the timeout may be reached. 
On the other hand, if the data storage being queried may also experience jitters, such as HDFS slow nodes, this will also cause intermittent lags in the query.

**Comment by njalan** [comment]
I have multiple clusters with different nodes (5-20) and each node is 16 core and 128G.  Concurrency is 40 but the query is timeout when the cluster is not busy.  We are using s3 as the storage. Trino queries are called by third party app. So whether third party app's workload will also cause the query timeout?

**Reply by Heltman** [reply]
In our scenario, external systems will affect the query timeout, such as busy HDFS or S3 network problems, which will cause the query time to increase significantly. 
You said below that it is the timeout of the finishing state. However, in general, finishing state indicates that the query has been completed and no timeout error occurs.
You can observe the webui homepage to see if there are only 1-2 tasks executing for long queries. This indicates that there are slow nodes, which may be caused by underlying storage problems.

![image](https://github.com/user-attachments/assets/d3aba03c-3fd1-4288-91f5-b650aeacb35e)

**Comment by njalan** [comment]
@Heltman By the way most of the queries timeout are in **finishing state**

**Comment by njalan** [comment]
@Heltman  Is there anyway to find the root case of the timeout on finishing state. I need the evidence whether it is storage or work cpu issues.
